lib/full-site-editing/block-templates.php

```diff
@@ -386,41 +386,43 @@ function gutenberg_get_block_template( $id, $template_type = 'wp_template' ) {
  * @param string $post_type     Post type.
  * @return string The original, desired slug.
  */
-function gutenberg_filter_wp_template_unique_post_slug( $slug, $post_ID, $post_status, $post_type ) {
-	if ( 'wp_template' !== $post_type || 'wp_template_part' !== $post_type ) {
+function gutenberg_filter_wp_template_unique_post_slug( $placeholder, $slug, $post_ID, $post_status, $post_type ) {
+	if ( 'wp_template' !== $post_type && 'wp_template_part' !== $post_type ) {
 		return $slug;
 	}
 
 	// Template slugs must be unique within the same theme.
-	$theme = get_the_terms( $post_ID, 'wp_theme' )[0]->slug;
+	$theme = wp_get_theme()->get_stylesheet();
 
 	$check_query_args = array(
-		'post_name'      => $slug,
+		'post_name__in'  => array( $slug ),
 		'post_type'      => $post_type,
 		'posts_per_page' => 1,
+		'no_found_rows'  => true,
 		'post__not_in'   => $post_ID,
 		'tax_query'      => array(
-			'taxonomy' => 'wp_theme',
-			'field'    => 'name',
-			'terms'    => $theme,
+			array(
+				'taxonomy' => 'wp_theme',
+				'field'    => 'name',
+				'terms'    => $theme,
+			),
 		),
-		'no_found_rows'  => true,
 	);
 	$check_query      = new WP_Query( $check_query_args );
 	$posts            = $check_query->get_posts();
 
 	if ( count( $posts ) > 0 ) {
 		$suffix = 2;
 		do {
-			$query_args              = $check_query_args;
-			$alt_post_name           = _truncate_post_slug( $slug, 200 - ( strlen( $suffix ) + 1 ) ) . "-$suffix";
-			$query_args['post_name'] = $alt_post_name;
-			$query                   = new WP_Query( $check_query_args );
+			$query_args                  = $check_query_args;
+			$alt_post_name               = _truncate_post_slug( $slug, 200 - ( strlen( $suffix ) + 1 ) ) . "-$suffix";
+			$query_args['post_name__in'] = array( $alt_post_name );
+			$query                       = new WP_Query( $query_args );
 			$suffix++;
 		} while ( count( $query->get_posts() ) > 0 );
 		$slug = $alt_post_name;
 	}
 
 	return $slug;
 }
-add_filter( 'wp_unique_post_slug', 'gutenberg_filter_wp_template_unique_post_slug', 10, 4 );
+add_filter( 'pre_wp_unique_post_slug', 'gutenberg_filter_wp_template_unique_post_slug', 10, 5 );
```

Comment:
We need to do this on the `pre-*` hook.  Otherwise, it executes too late (even on 1st priority), and the slug it receives has already been appended with a suffix.